### PR TITLE
Remove memref runtime dependency and generate descriptors at compile time

### PR DESCRIPTION
**Pull request by asaadaldien**
At runtime buffers don't have shape or stride information to construct memref descriptors, so the runtime should only pass  raw_ptr & push_constant and the generated code must generate well formed descriptors.

This is done as the following: 
- Entry point invocation ABI for dylib and llvmjit runtime `void foo(void** packed_buffer_args, int64_t*  push_constant)`
- LLVMIR entry point function generated `func foo(%packed_buffer_args: !<i8**>,  %push_constant: !llvm<i64*>)`
- Delay the construction of entry function signature to final LLVM lowering step.
- A pattern that lowers `iree.placeholder` and hal.interface.load_constant` to LLVMIR.
- Replace `iree.placeholder` ops with inferred memref descriptors.
- Replace `hal.interface.load_constant` with `llvm.load`
- Remove HALInterfaceToMemrefArguments pass.
- Remove memre_runtime.h
- Remove dead code related to `inovke_` and `emit_c_interface`

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp [hunk 1]
```diff
@@ -124,6 +124,182 @@ class RemoveMakeRankedShape : public ConvertToLLVMPattern {
...
+  if (!type) return nullptr;
+  auto *mlirContext = type.getContext();
+  auto wrappedLLVMType = type.dyn_cast<LLVM::LLVMType>();
+  if (!wrappedLLVMType)
+    emitError(UnknownLoc::get(mlirContext),
+              "conversion resulted in a non-LLVM type");
+  return wrappedLLVMType;
+}
+
+class ConvertFuncWithHALInterFace : public ConvertToLLVMPattern {
```
Comment (benvanik) [inline iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp:147]: Interface*

File: iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp [hunk 2]
```diff
@@ -124,6 +124,182 @@ class RemoveMakeRankedShape : public ConvertToLLVMPattern {
...
+                             typeconverter) {}
+
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    if (SymbolTable::getSymbolVisibility(op) != SymbolTable::Visibility::Public)
+      return failure();
+    auto funcOp = dyn_cast_or_null<FuncOp>(op);
+    FunctionType fnType = funcOp.getType();
+    if (fnType.getNumInputs() != 0)
```
Comment (benvanik) [inline iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp:161]: style nit: use {} around multi-line if statements

File: iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp [hunk 3]
```diff
@@ -124,6 +124,182 @@ class RemoveMakeRankedShape : public ConvertToLLVMPattern {
...
+          elementType.getPointerTo(memrefType.getMemorySpace()));
+    }
+
+    TypeConverter::SignatureConversion signatureConverter(/*numOrigInputs=*/0);
+
+    // func foo(%buffer_args: !llvm<i8**>, %dim_valus: !llvm<i32*>)
+    auto interfaceInputArg =
+        LLVM::LLVMType::getInt8PtrTy(typeConverter.getDialect()).getPointerTo();
+    auto dimValuesArgType =
+        LLVM::LLVMType::getInt64Ty(typeConverter.getDialect()).getPointerTo();
```
Comment (benvanik) [inline iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp:226]: int64 doesn't match comment above (i32)?

File: iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp (hunk 1 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp]: Done!

File: iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp (hunk 2 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp]: Done!

File: iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp (hunk 3 above)
Comment (asaadaldien) [reply iree/compiler/Conversion/LinalgToLLVM/ConvertToLLVM.cpp:222]: Fixed the comment, the runtime push to vector<int64_t> there is a todo to consider move the casting to LLVMIR.

**Pull request merged**